Exclude-Generated-Columns-In-Copy

emelsimsek opened the pull request:
DESCRIPTION: Fixes a bug in shard copy operations.

For copying shards in both shard move and shard split operations, Citus uses the COPY statement.  

A COPY all statement in the following form
` COPY target_shard FROM STDIN;`
throws an error when there is a GENERATED column in the shard table.

In order to fix this issue, we need to exclude the GENERATED columns in the COPY and the matching SELECT statements. Hence this fix converts the COPY and SELECT all statements to the following form:
```
COPY target_shard (col1, col2, ..., coln) FROM STDIN;
SELECT (col1, col2, ..., coln) FROM source_shard;
```
where (col1, col2, ..., coln) does not include a GENERATED column. 
GENERATED column values are created in the target_shard as the values are inserted.

Fixes #6705.

marcocitus commented:
I think the problem is that we should not add GENERATED properties to the sharda to begin with, only to shell tables. The DDL propagation should be updated.

marcocitus reviewed (dismissed):
See above

JelteF reviewed (commented):
Should include some tests to make sure we don't regress in the future.

marcocitus commented:
Looking at it more closely, I guess GENERATED ALWAYS properties have to be pushed down to workers because they can change in case of update, and we would miss those changes unless we explicitly add assignments for those fields in the UPDATE.

We may have to be a little bit careful what we allow on the DDL side though, e.g. things can change per node:
```sql
create table test (x int, y int);
select create_distributed_table('test','x');
insert into test select generate_series(1,10) s;
alter table test add column n oid generated always as (tableoid) stored;
select * from test order by 1,2;
┌────┬───┬────────┐
│ x  │ y │   n    │
├────┼───┼────────┤
│  1 │   │ 242995 │
│  2 │   │ 255643 │
│  3 │   │ 243016 │
│  4 │   │ 255619 │
│  5 │   │ 255616 │
│  6 │   │ 255637 │
│  7 │   │ 255619 │
│  8 │   │ 255607 │
│  9 │   │ 255649 │
│ 10 │   │ 255613 │
└────┴───┴────────┘
(10 rows)
```
(though there's probably no good way to support this one, so maybe using the shard OID is an acceptable trade-off)

JelteF commented on src/backend/distributed/operations/worker_shard_copy.c:
```diff
@@ -344,21 +345,99 @@ ShardCopyDestReceiverDestroy(DestReceiver *dest)
+ConstructTargetColumnsList(TupleDesc tupdesc)
+{
+	StringInfo columnList = makeStringInfo();
+	int i;
+
+	for (i = 0; i < tupdesc->natts; i++)
```
nit:

```suggestion
	for (int i = 0; i < tupdesc->natts; i++)
```

emelsimsek replied:
Done.

JelteF commented on src/backend/distributed/operations/worker_shard_copy.c:
```diff
@@ -344,21 +345,99 @@ ShardCopyDestReceiverDestroy(DestReceiver *dest)
+
+	bool first = true;
+
+	for (int attrNum = 1; attrNum <= numberOfAttributes; attrNum++)
+	{
+		Form_pg_attribute attributeTuple =
+			TupleDescAttr(relation->rd_att, attrNum - 1);
+
+		if (attributeTuple->attgenerated || attributeTuple->attisdropped)
+		{
+			continue;
+		}
+
+		if (!first)
+		{
+			appendStringInfo(columnList, ", ");
+		}
+
+		appendStringInfo(columnList, "%s ", NameStr(attributeTuple->attname));
+
+		first = false;
+	}
+
```
Instead of implementing the same loop twice, I think this can call `ConstructTargetColumnsList(relation->rd_att)`

emelsimsek commented:
@JelteF, this change is ready for a re-review I believe.

JelteF commented on src/backend/distributed/operations/worker_shard_copy.c:
```diff
@@ -344,21 +345,80 @@ ShardCopyDestReceiverDestroy(DestReceiver *dest)
+ *  statement fails to handle them. Iterating over the attributes of the table we also need to skip the dropped columns.
+ */
+const char *
+GenerateColumnListFromTupleDesc(TupleDesc tupDesc)
```
```suggestion
CopyableColumnNamesFromTupleDesc(TupleDesc tupDesc)
```

JelteF commented on src/backend/distributed/operations/worker_shard_copy.c:
```diff
@@ -344,21 +345,80 @@ ShardCopyDestReceiverDestroy(DestReceiver *dest)
+ *  GenerateColumnListFromRelationName function is a wrapper for GenerateColumnListFromTupleDesc.
+ */
+const char *
+GenerateColumnListFromRelationName(const char *relationName, const char *schemaName)
```
Previous name is not very descriptive and generally schemaname comes before tablename.
```suggestion
CopyableColumnNamesFromRelationName(const char *schemaName, const char *relationName)
```

JelteF commented on src/test/regress/sql/worker_split_copy_test.sql:
```diff
@@ -110,8 +110,61 @@ SELECT COUNT(*) FROM worker_split_copy_test."test !/ \n _""dist_123_table_810700
+-- BEGIN: Create distributed table and insert data.
+CREATE TABLE worker_split_copy_test.dist_table_with_generated_col(id int primary key, new_id int GENERATED ALWAYS AS ( id + 3 ) stored, value char);
+SELECT create_distributed_table('dist_table_with_generated_col', 'id');
+INSERT INTO dist_table_with_generated_col (id, value) (SELECT g.id, 'N' FROM generate_series(1, 1000) AS g(id));
```
Would be good to drop the value column here too. To also test for dropped columns, just like is done for shard moves.

JelteF commented on src/test/regress/sql/worker_split_copy_test.sql:
```diff
@@ -110,8 +110,61 @@ SELECT COUNT(*) FROM worker_split_copy_test."test !/ \n _""dist_123_table_810700
+SELECT COUNT(*) FROM worker_split_copy_test.dist_table_with_generated_col_81080016;
+
+\c - - - :worker_1_port
+SELECT * from worker_split_copy(
```
This is testing the low level function only. I think it would be good to add a generated column to the table in `citus_split_shard_by_split_points.sql` and `citus_non_blocking_split_shards.sql`to also test everything end-to-end.

JelteF reviewed (dismissed):
added some final comments for regression test additions. Other than that this looks good.

emelsimsek commented:
> added some final comments for regression test additions. Other than that this looks good.

@JelteF, I believe I have addressed all the review comments now. But looks like I need another approval.
